modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,439 @@
+	}
+
+	/**
+	 * Outputs the contents of the widget on the front end.
+	 *
+	 * If the widget is unconfigured, a configuration message is displayed to users with admin access
+	 * and the entire widget is hidden from everyone else to avoid displaying an empty widget.
+	 *
+	 * @param array $args The sidebar arguments that control the wrapping HTML.
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function widget( $args, $instance ) {
+		$instance = wp_parse_args( $instance, $this->defaults );
+
+		// Don't display anything to non-blog admins if the widgets is unconfigured
+		if ( ! $instance['token_id'] && ! current_user_can( 'edit_theme_options' ) )
+			return;
+
+		echo $args['before_widget'];
+
+		// Always show a title on an unconfigured widget
+		if ( ! $instance['token_id'] && empty( $instance['title'] ) )
+			$instance['title'] = $this->defaults['title'];
+
+		if ( ! empty( $instance['title'] ) )
+			echo $args['before_title'] . esc_html( $instance['title'] ) . $args['after_title'];
+
+		if ( ! $instance['token_id'] ) {
+			echo 'This widget cannot make new connections to Instagram. You can install and use a third-party Instagram plugin instead. Please <a href="https://wordpress.com/help/contact/">contact us if you need help</a> setting this up.';
+			echo '<p><em>This notice is only shown to you.</em></p>';
+			echo '<a class="button-primary" target="_top" href="https://public-api.wordpress.com/connect/?action=request&amp;kr_nonce=25051df3f2&amp;nonce=84e239e897&amp;for=instagram-widget&amp;service=instagram&amp;blog=105567107&amp;kr_blog_nonce=217ce9172b&amp;magic=keyring&amp;instagram_widget_id=8">Authorize Instagram Access</a>';
+		} else {
+			$images = $this->get_images( $instance );
+
+			if ( ! is_array( $images ) ) {
+				echo '<p>' . __( 'There was an error retrieving images from Instagram. An attempt will be remade in a few minutes.', 'wpcomsh' ) . '</p>';
+			}
+			elseif ( ! $images ) {
+				echo '<p>' . __( 'No Instagram images were found.', 'wpcomsh' ) . '</p>';
+			} else {
+
+				echo '<div class="' . esc_attr( 'wpcom-instagram-images wpcom-instagram-columns-' . (int) $instance['columns'] ) . '">' . "\n";
+				foreach ( $images as $image ) {
+					echo '<a href="' . esc_url( $image['link'] ) . '" target="' . esc_attr( apply_filters( 'wpcom_instagram_widget_target', '_self' ) ) . '"><div class="sq-bg-image" style="background-image: url(' . esc_url( set_url_scheme( $image['url'] ) ) . ')"><span class="screen-reader-text">' . esc_attr( $image['title'] ) . '</span></div></a>' . "\n";
+				}
+				echo "</div>\n";
+			}
+		}
+
+		echo $args['after_widget'];
+	}
+
+	/**
+	 * Outputs the widget configuration form for the widget administration page.
+	 * Allows the user to add new Instagram Keyring tokens and more.
+	 *
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function form( $instance ) {
```

Comment:
This is essentially a copy/paste from the widget in D40045, but with some changes for Jetpack